rfcs/control-groups/README.md

```diff
@@ -0,0 +1,279 @@
+
+![cgroupv2 hierarchy](./cgroupv2-hierarchy.jpg)
+
+
+### What we have
+
+ClickHouse already has things similar to cgroup controllers that are used to
+implement various limits.
+
+Examples: [MemoryTracker][MemoryTracker.h], Network [Throttler][Throttler.h].
+
+[ProfileEvents][ProfileEvents.h] and thread groups hierarchies with
+[ThreadProfileEvents][ThreadProfileEvents.h] could fall in this category as well
+even though today only tracking is supported (rather than "controlling").
+[3] has some neat ideas on how it could be possible to "bind" actions to events.
+
+An important property to mention of the above mentioned "things" is that all of
+them are hierarchical.
+
+The missing part is that there isn't a lot of flexibility in configuring these.
+
+
+### A rough sketch of how the future could look like
+
+Let's draw some inspiration from Linux control groups and adapt the existing
+"resource controllers" to achieve "resource pools".
+
+* There should be a root control group for the system limits
+
+* It should be possible to define control groups for background processes
+and assign tables to these control groups.
+    * Ie a table or a group of tables could use at most `X b/s` of disk, 
+    `Y b/s` of network for replication, `Z b` of storage.
+        * We could instead use writer's control group but there are things like
+        Kafka engine which would need some special handling.
+
+* It should be possible to define control groups for queries to limit eg CPU,
+query concurrency, max memory usage.
+    * This could help isolate writing vs reading.
+
+* It should be possible to assign users/roles to control groups.
+    * Maybe even multiple ones and allow users to choose which control groups
+    should be used for operations (similar to roles today?).
+
+* It should be possible to nest control groups to form hierarchies.
+
+* It should be possible to configure per-user and per-query limits using control
+groups in order to achieve "consistent implementation" of the limits.
+    * Maybe allow something like `CREATE CONTROL GROUP TEMPLATE zzz FOR QUERY ...`,
+    `CREATE CONTROL GROUP TEMPLATE zzz FOR USER` (`... FOR QUOTA KEY`?).
+        * These would be applied for each query/user separately
+        and automatically "nest" under the control group to which the user belongs.
+
+
+### First rocks in the road
+
+It is not very clear what the implementation of "(resource) control group" should look like.
+Eg the global control group needs to know about query concurrency limits, memory limits.
+Query concurrency limits in control groups for background processes hardly makes sense.
+
+An idea would be to have "control groups" to which controllers like memory tracker,
+network throttler, query concurrency (yes, make it a controller) could be attached
+and configured independently.
+
+Configuration could be inspired by [3] where for hitting limits is accompanied
+by actions like "queue/wait", "throw", "evict lower priority".
```

Comment:
We need to think how we see common workload. Currently we fire query with maximum parallelism if there is no limit. It will try to run ASAP and hit limit to stop receiving more resources? Should it be the common case? Probably we need to design query control flow in which it gets information from controllers to set optimal parallelism etc.
I have a point that if isolation works, user should not use max_threads/max_memory_usage etc for his queries. We should consider that they add only information about in which group this query should run.